### DISCUSSION TITLE
How I started using odmantic

### DISCUSSION THREAD

**Discussion by aahnik** [art049/odmantic#144] [Show and tell]
I am an 18 yo beginner developer from India. I am building a "refer and earn" telegram bot for my friend. (currently a private project, but plan to open source it sometime in future)

I find it more natural and to think in terms of python objects. Pydantic is just awesome as it does everything so elegantly. 

For rapid prototyping, I had just used an in-memory python dictionary with the user id as key, and the Person model as the value.

Previously my code was something like this in `storage.py`

```python
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel

class Person(BaseModel):
    uid: int
    referer: Optional[int] = None
    joined: StrictBool = False
    referals: Optional[List[int]] = []
    coins: int = 0
    wallet: str = ""


stored: Dict[int, Person] = {}
```

In other modules to access or manipulate the data, I did something like this

```python
from myproject import storage as st

# create a new person
st.stored[some_id] = st.Person(uid=some_id)

# get an existing person
person = st.stored.get(some_id)

# update existing person
person.coins = 1
st.stored[person.uid] = person
```

-----------------

What I did was working, but I needed two things

- Persistence
- Scalability

I did not like the idea of using SQL, as I would be rapidly changing the structure of data in the future. I had never used MongoDB at this point. But I thought let me give it a shot. (I don't know much of theoretical stuff, just exploring!) Another reason I did not use SQL is that I found SQLAlchemy a little scary.

While  I was searching the web and exploring GitHub, **I found the magic wand! odmantic.**

_Odmantic made it absolutely easy for me to switch from my rough prototype to a serious and powerful MongoDB database._

Switching to `odmantic` was easy.

```diff
- from pydantic import BaseModel
+ from odmantic import AIOEngine, Field, Model

- class Person(BaseModel):
+ class Person(Model):

-    uid: int
+    uid: int = Field(primary_field=True)

- stored: Dict[int, Person] = {}
+ engine: AIOEngine = None
```

At the entry point where I started my bot client, I also initialized my AIOEngine

```python
# bot.py
from myproject import storage as st

engine = AIOEngine()
st.engine = engine
```

Now in other modules to access or manipulate the data, I simply do like this.

```diff
from myproject import storage as st

# create a new person
- st.stored[some_id] = st.Person(uid=some_id)
+ person = st.Person(uid=some_id)
+ await st.engine.save(person)

# get an existing person
- person = st.stored.get(some_id)
+  person: st.Person = await st.engine.find_one(
+            st.Person, st.Person.uid == some_id
 +       )

# update existing person
person.coins = 1
- st.stored[person.uid] = person
+ await st.engine.save(person)
```

From now on I think I will be using MongoDB with odmantic for my future projects. It's much easy to define the data structure, and everything is fast and fully scalable.

Thanks, @art049 for creating odmantic! 

**As of June 26th, 2021, the bot has nearly 5K users and everything seems to be working fine**. I also did schema changes by adding new attributes to the Person model. As I had set a default value for the new fields, there was no issue.

Anyone reading this feel free to suggest improvements or comment about your experiences.

**Comment by chrisdewa** [comment]
I agree! odmantic is truly fantastic! I use it mostly for discord bots and it's a breeze.
I might as well make a show and tell later :)
